### PR TITLE
fix `Host` and `:authority` header not being updated on changes to host or port

### LINKED ISSUE #5907: Map remote should include port in Host header (mitmweb)
#### Problem Description

When using `--map-remote` in `mitmweb`, the request URL gets replaced, and the `Host` gets updated. However, the port is not added to the Host header, leading to problems when using non-standard ports.

#### Steps to reproduce the behavior:

1. `mitmweb --map-remote '|http://[^/]+|http://portquiz.net:4444' --listen-port 9090'`
2. `curl --proxy http://127.0.0.1:9090 http://github.com`
3. Host header does not contain port
4. It should: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host

#### System Information

```
Mitmproxy: 9.0.1
Python:    3.11.1
OpenSSL:   OpenSSL 1.1.1s  1 Nov 2022
Platform:  macOS-13.2-arm64-arm-64bit
``

### PR DESCRIPTION
**Pull request by sujaldev**
#### Description

Fixes #5907

#### Checklist

 - [x] I have updated tests where applicable.
 - [ ] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Comment by sujaldev** [comment]
Seems like I screwed up, trying to fix it :(

**Review by mhils — CHANGES_REQUESTED** [review]
This requires significantly more thought. We want to update the host header and `.authority` both on updates to `.host` as well as on updates to `.port`. This also requires extensive testing. For example, the current PR will fail for `req.port = 42; req.port = 44`

File: mitmproxy/http.py [hunk 1]
```diff
@@ -795,6 +789,22 @@ def port(self) -> int:
     @port.setter
     def port(self, port: int) -> None:
         self.data.port = port
+        self._update_host_and_authority()
+
+    def _update_host_and_authority(self) -> None:
+        val = url.hostport(
+            self.scheme,
+            self.host,
+            # test_http.py::TestRequestCore::test_port sets port = b"foo" as the port
```
Comment (mhils) [inline mitmproxy/http.py:798]: Let's check that `port` is an `int` in `port.setter` and raise a ValueError if that's not the case. We can then adjust the test and skip this check here :)

**Review by mhils — COMMENTED** [review]
Nice, this looks much better now! 😃

**Review by mhils — APPROVED** [review]

**Pull request merged**